base/abstractset.jl

```diff
@@ -69,9 +69,11 @@ function union!(s::AbstractSet, sets...)
 end
 
 max_values(::Type) = typemax(Int)
-max_values(T::Type{<:Union{Nothing,BitIntegerSmall}}) = 1 << (8*sizeof(T))
-max_values(T::Union) = max(max_values(T.a), max_values(T.b))
+max_values(T::Union{map(X -> Type{X}, BitIntegerSmall_types)...}) = 1 << (8*sizeof(T))
+# saturated addition to prevent overflow with typemax(Int)
+max_values(T::Union) = max(max_values(T.a), max_values(T.b), max_values(T.a) + max_values(T.b))
 max_values(::Type{Bool}) = 2
+max_values(::Type{Nothing}) = 1
```

Comment:
Could this be applied to all singleton types? You can check whether a type is a singleton using `isdefined(T, :instance)` (there's an `issingletontype` helper function for this in replace.jl).

Reply:
Yes seems like a good idea!
As CI is green and I think this is an urgent fix to have, I will merge as-is later today and implement your idea (which is not directly related to the bugfix) separately.